src/Components/Components/src/PersistentState/PersistentServicesRegistry.cs

```diff
@@ -0,0 +1,230 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.AspNetCore.Components.Reflection;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Components.Infrastructure;
+
+internal sealed class PersistentServicesRegistry
+{
+    private static readonly string _registryKey = typeof(PersistentServicesRegistry).FullName!;
+
+    private readonly IServiceProvider _serviceProvider;
+    private readonly PersistentServiceTypeCache _persistentServiceTypeCache;
+    private IPersistentComponentRegistration[] _registrations;
+    private List<PersistingComponentStateSubscription> _subscriptions = [];
+    private static readonly ConcurrentDictionary<Type, PropertiesAccessor> _cachedAccessorsByType = new();
+
+    public PersistentServicesRegistry(IServiceProvider serviceProvider)
+    {
+        var registrations = serviceProvider.GetRequiredService<IEnumerable<IPersistentComponentRegistration>>();
+        _serviceProvider = serviceProvider;
+        _persistentServiceTypeCache = new PersistentServiceTypeCache();
```

Comment:
Since this registry is a scoped service, is it worth having the cache scoped as well, shouldn't it be a singleton?

Reply:
Yeah, good point. In the beginning PersistentServicesRegistry was a singleton in DI, but had to change it for DI scoping reasons, so it makes sense for this to be a singleton/static instance. I'll change it.

Reply:
I've made the cache a static readonly, which is what it should be.